swarm/network/fetcher.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+var searchTimeout = 3000 * time.Millisecond
+
+// Fetcher is created when a chunk is not found locally
+// it starts a request handler loop once and keeps it
+// alive until all active requests complete
+// either because the chunk is delivered or requestor cancelled/timed out
+// fetcher self destroys
+// TODO: cancel all forward requests after termination
+type Fetcher struct {
+	request   func(context.Context, storage.Address, storage.Address, bool, *sync.Map) (context.Context, error) //
+	addr      storage.Address                                                                                   //
+	offerC    chan storage.Address
+	skipCheck bool
+}
+
+func NewFetchFunc(request func(context.Context, storage.Address, storage.Address, bool, *sync.Map) (context.Context, error), skipCheck bool) storage.FetchFuncConstructor {
```

Comment:
Defining:

```
type RequestFunc func(context.Context, storage.Address, storage.Address, bool, *sync.Map) (context.Context, error)
```

And using it in Fetcher struct and NewFetchFunc and NewFetcher functions would make this part of code a bit easier to read.

Reply:
as @janos said yesterday - this might be a place for a factory pattern, or, in my eyes, for a producer-consumer pattern implementation that would reduce complexity (and would allow to proxy/pool resources).
the original `request` param name suggests this should be rather a stateful object than a function pointer, which will then be passed down to a factory/producer component